### PR TITLE
Add support for types which implement encoding.TextUnmarshaler

### PR DESCRIPTION
**Pull request by albrow**
This allows the go-envvar package to support common types in the standard library such as time.Time. It also allows support for custom types, provided they implement the interface.

### CODE REVIEW

File: envvar/envvar.go [hunk 1]
```diff
@@ -137,6 +141,42 @@ func getStructTag(field reflect.StructField, key string) (value string, found bo
 // setFieldVal first converts v to the type of structField, then uses reflection
 // to set the field to the converted value.
 func setFieldVal(structField reflect.Value, name string, v string) error {
+
+	// Check if the struct field type implements the encoding.TextUnmarshaler
+	// interface.
+	if structField.Type().Implements(reflect.TypeOf([]encoding.TextUnmarshaler{}).Elem()) {
```
Comment (ntindall) [inline envvar/envvar.go]: I don't believe this code path is currently tested. Moreover, I think it should be required that the `TextUnmarshaler` interface be implemented with a pointer receiver, otherwise the holder would not be mutated when `UnmarshalText` was called.

Am I missing something @albrow? I think this code path should be an error.

File: envvar/envvar.go [hunk 2]
```diff
@@ -137,6 +141,42 @@ func getStructTag(field reflect.StructField, key string) (value string, found bo
...
+		// value. This should always be true, because the Parse method is strict
+		// about accepting a pointer to a struct type. However, if it's not true the
+		// Addr() call will panic, so it is good practice to leave this check in
+		// place. (In the reflect package, a struct field is considered addressable
+		// if we originally received a pointer to the struct type).
+		if structField.CanAddr() {
+			results := structField.Addr().MethodByName("UnmarshalText").Call([]reflect.Value{reflect.ValueOf([]byte(v))})
+			if !results[0].IsNil() {
+				err := results[0].Interface().(error)
+				return fmt.Errorf("envvar: Error parsing environment variable %s: %s\n%s", name, v, err.Error())
```
Comment (ntindall) [inline envvar/envvar.go]: can we add a test for this new error condition as well?

File: envvar/envvar.go (hunk 1 above)
Comment (albrow) [reply envvar/envvar.go]: Technically, it is not impossible for a type to correctly implement `TextUnmarshaler` with a non-pointer receiver for the `UnmarshalText` method. If you have a reference type within the receiver type, it will be settable even if the receiver itself is not. I added a test case to illustrate this:

```go
// customUnmarshallerWrapper also implements the UnmarshalText method by calling
// it on its own *customUnmarshaller.
type customUnmarshallerWrapper struct {
	um *customUnmarshaller
}

// UnmarshalText simply calls um.UnmarshalText. Note that here we use a
// non-pointer receiver. It still works because the um field is a pointer. We
// just need to be sure to check if um is nil first.
func (cuw customUnmarshallerWrapper) UnmarshalText(text []byte) error {
	if cuw.um == nil {
		return nil
	}
	return cuw.um.UnmarshalText(text)
}
```

It doesn't have to a pointer, either. It could also be a type whose underlying type is a pointer (e.g. `type myCustomType *int`) or a builtin reference type like `map`.

Honestly though, after looking at this again, I'm not sure how often it will occur in the real world. I would be open to hearing an argument that this edge case is not necessary and should be removed.

File: envvar/envvar.go (hunk 2 above)
Comment (albrow) [reply envvar/envvar.go]: 👌 added.

**Comment by albrow** [comment]
I also added some test cases for setting default values on a field which implements `TextUnmarshaler`.

File: envvar/envvar_test.go [hunk 3]
```diff
@@ -40,6 +48,9 @@ func TestParse(t *testing.T) {
 		FLOAT32: 0.001234,
 		FLOAT64: 23.7,
 		BOOL:    true,
+		TIME:    time.Date(2017, 10, 31, 14, 18, 0, 0, time.UTC),
+		CUSTOM:  customUnmarshaler{strings: []string{"foo", "bar", "baz"}},
```
Comment (ntindall) [inline envvar/envvar_test.go]: nit: inconsistent formatting with [line 92](https://github.com/plaid/go-envvar/pull/9/files#diff-e6feeb46d21c27c75be297aee773211cR92)

File: envvar/envvar_test.go [hunk 4]
```diff
@@ -141,6 +157,50 @@ func TestParseWithInvalidArgs(t *testing.T) {
...
+	cu.strings = strings.Split(string(text), ",")
+	return nil
+}
+
+// customUnmarshalerWrapper also implements the UnmarshalText method by calling
+// it on its own *customUnmarshaler.
+type customUnmarshalerWrapper [3]string
+
+// UnmarshalText simply calls um.UnmarshalText. Note that here we use a
+// non-pointer receiver. It still works because the um field is a pointer. We
```
Comment (ntindall) [inline envvar/envvar_test.go]: comment seems out of date now, (i think this is a better example than we have in the discussion, though).

File: envvar/envvar_test.go [hunk 5]
```diff
@@ -141,6 +157,50 @@ func TestParseWithInvalidArgs(t *testing.T) {
...
+		cuw[i] = values[i]
+	}
+	return nil
+}
+
+// errorUnmarshaler implements the UnmarshalText method by always returning
+// an error.
+type alwaysErrorUnmarshaler struct{}
+
+func (eu alwaysErrorUnmarshaler) UnmarshalText(text []byte) error {
```
Comment (ntindall) [inline envvar/envvar_test.go]: Maybe its pedantic, but should we test the case for the pointer receiver as well since that goes down a different code path?

File: envvar/envvar.go (hunk 1 above)
Comment (ntindall) [reply envvar/envvar.go]: I see. I hadn't considered this...

While the example you have here with `customUnmarshallerWrapper` seems a little unlikely to come up, I think think unmarshalling to a builtin reference type is something we would want to support. 👍

File: envvar/envvar_test.go [hunk 6]
```diff
@@ -141,6 +157,50 @@ func TestParseWithInvalidArgs(t *testing.T) {
...
+}
+
+// UnmarshalText simply splits the text by the separator: ",".
+func (cu *customUnmarshaler) UnmarshalText(text []byte) error {
+	cu.strings = strings.Split(string(text), ",")
+	return nil
+}
+
+// customUnmarshalerWrapper also implements the UnmarshalText method by calling
+// it on its own *customUnmarshaler.
```
Comment (ntindall) [inline envvar/envvar_test.go]: comment is out of date now.

File: envvar/envvar_test.go (hunk 4 above)
Comment (albrow) [reply envvar/envvar_test.go]: I'm sorry, there were actually some unintended changes included in one of the recent commits. That is why some of the comments don't line up with the code. I agree that the version of `customUnmarshalerWrapper` which used a slice type (`[]string`) is easier to understand, but unfortunately it doesn't work. You can't set the value of a slice the way that we would like to.

I've fixed the PR so that now only the intended changes are included, and the comments are re-aligned with the code.

(Technically the built-in slice type contains a pointer to an underlying array. While you can change that array, the changes won't necessarily be reflected in any slices that point to the same array. IMO this idiosyncrasy will lead to more confusion than just using a pointer in our examples and tests).

File: envvar/envvar_test.go (hunk 3 above)
Comment (albrow) [reply envvar/envvar_test.go]: ⚡️

File: envvar/envvar_test.go (hunk 6 above)
Comment (albrow) [reply envvar/envvar_test.go]: ⚡️

**Comment by albrow** [comment]
@ntindall I believe I have addressed all of your comments. Should be ready for final review.

**Comment by albrow** [comment]
Thanks @ntindall. Looks like I don't have the ability to merge, so I'll leave that to you.

**Comment by ntindall** [comment]
@albrow could you please rebase this against master ?

**Comment by ntindall** [comment]
LGTM.

Thanks @albrow !
